abgc/src/lib.rs

```diff
@@ -30,75 +29,101 @@ pub trait GcLayout {
 
 #[derive(Debug)]
 pub struct Gc<T: GcLayout> {
-    ptr: *mut GcBox<T>,
+    objptr: *mut T,
 }
 
 impl<T: GcLayout> Gc<T> {
     /// Create a `Gc` from `v`.
     pub fn new(v: T) -> Self {
-        let (gcbptr, vptr): (_, *mut T) = GcBox::<T>::alloc_blank(Layout::new::<T>());
+        let objptr = Gc::<T>::alloc_blank(Layout::new::<T>());
         let gc = unsafe {
-            vptr.copy_from_nonoverlapping(&v, 1);
-            Gc::from_raw(gcbptr)
+            objptr.copy_from_nonoverlapping(&v, 1);
+            Gc::from_raw(objptr)
         };
         forget(v);
         gc
     }
 
+    /// Allocate memory sufficient to `l` (i.e. correctly aligned and of at least the required
+    /// size). The returned pointer must be passed to `Gc::from_raw`.
+    pub fn alloc_blank(l: Layout) -> *mut T {
+        let (layout, uoff) = Layout::new::<usize>().extend(l).unwrap();
+        // In order for our storage scheme to work, it's necessary that `uoff - sizeof::<usize>()`
+        // gives a valid alignment for a `usize`. There are only two cases we need to consider
+        // here:
+        //   1) `object`'s alignment is smaller than or equal to `usize`. If so, no padding will be
+        //      added, at which point by definition `uoff - sizeof::<usize>()` will be exactly
+        //      equivalent to the start point of the layout.
+        //   2) `object`'s alignment is bigger than `usize`. Since alignment must be a power of
+        //      two, that means that we must by definition be adding at least one exact multiple of
+        //      `usize` bytes of padding.
+        // The assert below is thus paranoia writ large: it could only trigger if `Layout` started
+        // adding amounts of padding that directly contradict the documentation.
+        debug_assert_eq!(uoff % align_of::<usize>(), 0);
+
+        unsafe {
+            let baseptr = alloc(layout);
+            let objptr = baseptr.add(uoff);
+            let clonesptr = objptr.sub(size_of::<usize>());
```

Comment:
Couldn't you clone the original value of `baseptr` to avoid the subtraction?

Reply:
That really is silly of me!

Reply:
Ah, no, d'oh, sorry I mixed this up with some other code I had in another repository. This is actually correct, but it's very subtle. Imagine you want to store an item with alignment 1024 bytes: the layout we will use will then be (on a 64 bit machine):

  0: baseptr
  1016: clones
  1024: object

In other words, we just move the padding (of 1016 bytes) around because `Layout` told us to do:

  0: clones
  1024: object

Why do we do this? Because if we were to use the default `Layout`, we'd need to always know the alignment of `object` in order to work out where it's stored. Using our altered layout, we can keep a point to the beginning of the object and simply subtract `usize` bytes from it to find the `clones` count.

Reply:
I'm not sure I follow. Does this means that if we align to 8 bytes on a 64 bit machine, the size of the object would be `sizeof(T) + sizeof(usize)`, but for any layout with an alignment greater than 8 bytes, the size would simply be `sizeof(T)`, because the `clones` count was encoded in the spare padding bits?

Reply:
ok

Reply:
> Does this means that if we align to 8 bytes on a 64 bit machine, the size of the object would be sizeof(T) + sizeof(usize)

Yes.

> but for any layout with an alignment greater than 8 bytes, the size would simply be sizeof(T), because the clones count was encoded in the spare padding bits?

In practise, in such a case, the size would be `2 * sizeof(T)`.

Reply:
Ah yes, got it.

Reply:
In some ways it's a dumb trick, but since alignments greater than a `usize` are rare, it seems like a reasonable trade-off to me, at least for `abgc` in its current state.

Reply:
I think this is okay because it's not stack allocated. If the object was stack allocated the Sys V ABI specifies the stack must be 16 byte aligned on x64, which would mean that you would always use `2 * sizeof(T)`.

Reply:
Does every item in the stack have to be aligned to 16 bytes or just the stack pointer when a function is called? I assume the latter, but I'm not an expert.

Reply:
Ah yes, my bad, just the stack pointer. So this would be fine anyway.